app/src/main/scala/org/alephium/explorer/service/TokenSupplyService.scala

```diff
@@ -106,30 +143,40 @@ object TokenSupplyService {
       })
     }
 
-    private val mainInputs  = InputSchema.table.filter(_.mainChain)
-    private val mainOutputs = OutputSchema.table.filter(_.mainChain)
-
-    private def genesisLockedToken(lockTime: TimeStamp) = {
-      mainOutputs
-        .join(
-          mainTransactions
-            .join(BlockHeaderSchema.table.filter(_.height === Height.unsafe(0)))
-            .on(_.blockHash === _.hash)
-        )
-        .on(_.txHash === _._1.hash)
-        .filter { case (output, _) => output.lockTime.map(_ > lockTime).getOrElse(false) }
-        .map { case (output, _) => output.amount }
-        .sum
+    private def circulatingTokensQuery(at: TimeStamp) = {
+      sql"""
+      SELECT sum(CASE WHEN
+                     /* Only count unlock tokens */
+                     outputs.lock_time is NULL OR outputs.lock_time <= $at THEN outputs.amount
+                   ELSE 0
+                 END)
+      FROM outputs
+      LEFT JOIN inputs
+        ON outputs.key = inputs.output_ref_key
+        AND inputs.main_chain = true
+        AND inputs.block_timestamp <= $at
+      WHERE outputs.block_timestamp >= ${ALPH.LaunchTimestamp} /* We don't count genesis address */
+      AND outputs.block_timestamp <= $at
+      AND outputs.main_chain = true
+      AND outputs.spent_finalized IS NULL
+      AND outputs.address NOT IN #$excludedAddresses /* We exclude the reserved wallets */
+      AND inputs.block_hash IS NULL;
+        """.as[Option[U256]].exactlyOne
```

Comment:
Nice! I haven't tested this but since its summing a single column I'm wondering if the `case` condition (`outputs.lock_time is NULL OR outputs.lock_time <= $at THEN outputs.amount`) could be placed in the query's body so indexes get used filtering `lock_time`? Then in `SELECT` it could just be `sum(outputs.amount)`? 

Probably needs an index for `lock_time`.